src/index.js

```diff
@@ -0,0 +1,864 @@
+            // Have we seen this message before> if so, ignore
+            if(this.seenCache.has(seqno)){
+                return
+            }
+
+            this.seenCache.put(seqno)
+
+            // Emit to floodsub peers
+            // Need to figure this out
+
+            // Emit to peers in the mesh
+            let topics = msg.topicIDs
+            topics.forEach((topic) => {
+                let meshPeers = this.mesh.get(topic)
+                meshPeers.forEach((peer) => {
+                    if(!peer.isWritable || !utils.anyMatch(peer.topics, topics)) {
+                        return
+                    }
+
+                    peer.sendMessages(utils.normalizeOutRpcMessages([msg]))
+
+                    this.log('publish msgs on topics', topic, peer.info.id.toB58String())
+                })
+            })
+        })
+    }
+
+    /**
+     * Returns a buffer of a RPC message that contains a control message
+     *
+     * @param {Array<RPC.Message>} msgs
+     * @param {Array<RPC.ControlIHave>} ihave
+     * @param {Array<RPC.ControlIWant>} iwant
+     * @param {Array<RPC.ControlGraft>} graft
+     * @param {Array<RPC.Prune>} prune
+     *
+     * @returns {RPC Object}
+     *
+     */
+    _rpcWithControl (msgs, ihave, iwant, graft, prune) {
+        return {
+            msgs: msgs,
+            control: {
+                ihave: ihave,
+                iwant: iwant,
+                graft: graft,
+                prune: prune
+            }
+        }
+    }
+
+    /**
+     * Handles IHAVE messages
+     *
+     * @param {Peer} peer
+     * @param {RPC.controlMessage Object} controlRpc
+     *
+     * @returns {RPC.ControlIWant Object}
+     */
+    _handleIHave (peer, controlRpc) {
+        let iwant = new Set()
+
+        let ihaveMsgs = controlRpc.ihave
+        if(!ihaveMsgs) {
+            return
+        }
+
+        ihaveMsgs.forEach((msg) => {
+            let topic = msg.topicID
+
+            if (!this.mesh.has(topic)) {
+                return
+            }
+
+            let msgIDs = ihaveMsgs.messageIDs
+            msgIDs.forEach((msgID) => {
+                if (this.seenCache.has(msgID)) {
+                    return
+                }
+                iwant.add(msgID)
+            })
+        });
+
+        if (!iwant.length) {
+            return
+        }
+
+        this.log("IHAVE: Asking for %d messages from %s", iwant.length, peer.info.id.toB58String)
+        let iwantlst = []
+        iwant.forEach((msgID) => {
+            iwantlst.push(msgID)
+        })
+
+        return {
+            messageIDs: iwantlst
+        }
+    }
+
+    /**
+     * Handles IWANT messages
+     *
+     * @param {Peer} peer
+     * @param {RPC.control} controlRpc
+     *
+     * @returns {Array<RPC.Message>}
+     */
+    _handleIWant (peer, controlRpc) {
+        // @type {Map<string, RPC.Message>}
+        let ihave = new Map()
+
+        let iwantMsgs = controlRpc.iwant
+        if (!iwantMsgs){
```

Comment:
This comment is pretty subjective so feel free to ignore.
There are some cases of null checks and short-circuiting that I think are somewhat unnecessary if we assume that they are internal methods. Imo makes things trickier than need be.

In this case, if `controlRpc` is indeed a control object, it will always have an array `controlRpc.iwant` and any elements of the array will always have a property `messageIDs`. We would want to input guard for methods exposed to users, but if this is an internal method, we can do without the extra checks.

You might also consider always returning the same type of object in the short circuits, so you don't change the function return type (to `Array<RPC.Message> | undefined`). This will make processing the outputs less error-prone since the return type will always be the same.